### PR TITLE
resolved_ts: use smaller timeout when do check_leader

### LINKED ISSUE #15999: 1 store check_leader slow may affect may store to advance resolve_ts
### What version of TiKV are you using?

v6.5.5

### Steps to reproduce

- Run store-read workload.

- Use this [fail_point](https://github.com/tikv/tikv/commit/7a610a0cff5d311bccdc4264d4df4ab104a2a7d6) to inject 5s delay, delete this error injection after 15 seconds.

As you can see the following metrics, there are many TiKV affected and the max resolve_ts gap getting bigger.

image
image

### What did you expect?

Only 1 TiKV affected which has fail_point inject.

### What did happened?

Many TiKV affected.

### PR DESCRIPTION
**Pull request by crazycs520**
### What is changed and how it works?

Issue Number: Close #15999

What's Changed:
- use smaller timeout when do check_leader, to to reduce the impact of a single TiKV slowing down on the advance resolve_ts of other TiKVs

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

**Before**

image

image

**This PR**

image

image

Side effects

- N/A

### Release note 

```release-note
None.
```

### CODE REVIEW

**Review by you06 — APPROVED** [review]

File: components/resolved_ts/src/advance.rs [hunk 1]
```diff
@@ -275,7 +277,12 @@ impl LeadershipResolver {
     // This function broadcasts a special message to all stores, gets the leader id
     // of them to confirm whether current peer has a quorum which accepts its
     // leadership.
-    pub async fn resolve(&mut self, regions: Vec<u64>, min_ts: TimeStamp) -> Vec<u64> {
+    pub async fn resolve(
+        &mut self,
+        regions: Vec<u64>,
+        min_ts: TimeStamp,
+        timeout: Option<Duration>,
```
Comment (zyguan) [inline components/resolved_ts/src/advance.rs:284]: `timeout` is actually the timeout of inner grpc calls, in the worst case, `resolve` may take about `2*timeout`.

**Review by zyguan — APPROVED** [review]

File: components/resolved_ts/src/advance.rs (hunk 1 above)
Comment (you06) [reply components/resolved_ts/src/advance.rs:284]: Do you mean `get_tikv_client` takes a `timeout` duration but return `Ok<...>`, then the RPC takes another `timeout` duration? It's rare to be such worse IMO, but we may replace the `tokio::time::timeout` with `tokio::time::timeout_at` in the future.

Comment (zyguan) [reply components/resolved_ts/src/advance.rs:284]: Yes, I just put a comment here, the PR LGTM.

File: components/resolved_ts/src/advance.rs [hunk 2]
```diff
@@ -362,6 +369,10 @@ impl LeadershipResolver {
             .map_or(0, |req| req.regions[0].compute_size());
         let store_count = store_req_map.len();
         let mut check_leader_rpcs = Vec::with_capacity(store_req_map.len());
+        let timeout = match timeout {
+            Some(v) => cmp::min(DEFAULT_CHECK_LEADER_TIMEOUT_DURATION, v),
+            None => DEFAULT_CHECK_LEADER_TIMEOUT_DURATION,
+        };
```
Comment (cfzjywxk) [inline components/resolved_ts/src/advance.rs:375]:
```suggestion
        let timeout = timeout.unwrap_or(DEFAULT_CHECK_LEADER_TIMEOUT_DURATION)
        .min(DEFAULT_CHECK_LEADER_TIMEOUT_DURATION);

```

File: components/resolved_ts/src/advance.rs [hunk 3]
```diff
@@ -376,9 +387,16 @@ impl LeadershipResolver {
             let rpc = async move {
                 PENDING_CHECK_LEADER_REQ_COUNT.inc();
                 defer!(PENDING_CHECK_LEADER_REQ_COUNT.dec());
-                let client = get_tikv_client(to_store, pd_client, security_mgr, env, tikv_clients)
-                    .await
-                    .map_err(|e| (to_store, e.retryable(), format!("[get tikv client] {}", e)))?;
+                let client = get_tikv_client(
```
Comment (cfzjywxk) [inline components/resolved_ts/src/advance.rs:390]: Better to add a unit test case to verify the smaller timeout is used instead of the default one.

**Comment by LykxSassinator** [comment]
Should we also cherry-pick this pr to `release-6.5` ?

**Comment by cfzjywxk** [comment]
> Should we also cherry-pick this pr to `release-6.5` ?

I think so.
/cc @crazycs520 @zyguan

**Pull request merged**